definitions/scenarios/self_upgrade.rb

```diff
@@ -0,0 +1,96 @@
+module ForemanMaintain::Scenarios
+  class SelfUpgradeBase < ForemanMaintain::Scenario
+    def enabled_system_repos_id
+      feature(:system_repos).enabled_repos_ids
+    end
+
+    def enable_repos(repo_ids = stored_enabled_repos_ids)
+      add_step(Procedures::Repositories::Enable.new(repos: repo_ids))
+    end
+
+    def disable_repos(repo_ids = stored_enabled_repos_ids)
+      add_step(Procedures::Repositories::Disable.new(repos: repo_ids))
+    end
+
+    def target_version
+      @target_version ||= context.get(:target_version)
+    end
+
+    def current_version
+      feature(:instance).downstream.current_minor_version
+    end
+
+    def maintenance_repo(version)
+      "rhel-#{el_major_version}-server-satellite-maintenance-#{version}-rpms"
+    end
+
+    # Need to remove this before merging
+    def skip_repo_enablement?
+      !!context.get(:skip_repo_enablement)
+    end
+
+    def stored_enabled_repos_ids
+      unless defined?(@stored_enabled_repos_ids)
+        @stored_enabled_repos_ids = []
+        path = File.expand_path('enabled_repos.yml', ForemanMaintain.config.backup_dir)
+        @stored_enabled_repos_ids = File.file?(path) ? YAML.load(File.read(path)) : []
+      end
+      @stored_enabled_repos_ids
+    end
+
+    def all_maintenance_repos
+      search_id = "rhel-#{el_major_version}-server-satellite-maintenance-"
+      all_maintenance_repos = []
+      stored_enabled_repos_ids.each do |id|
+        next unless id.start_with?(search_id)
+        all_maintenance_repos << id
+      end
+      all_maintenance_repos
```

Comment:
I'd write it this way:
```suggestion
      repo_prefix = "rhel-#{el_major_version}-server-satellite-maintenance-"
      stored_enabled_repos_ids.filter { |id| id.start_with?(repo_prefix) }
```

You can even extract the `repo_prefix` definition to a separate method and reuse it in `def maintenance_repo()`. That said, you should consider that (AFAIK) `el_major_version` isn't cached and then it can become expensive.